src/bigm.jl

```diff
@@ -1,40 +1,68 @@
 """
-    big_m_reformulation!(constr::ConstraintRef, bin_var, M, i, j, k)
+    _calculate_tight_M
 
-Perform Big-M reformulation on a linear or quadratic constraint at index k of constraint j in disjunct i.
+Apply interval arithmetic on a linear constraint to infer the tightest Big-M value from the bounds on the constraint.
+"""
+function _calculate_tight_M(
+    con::JuMP.ScalarConstraint{JuMP.AffExpr, <: _MOI.LessThan}
+    )
+    return _interval_arithmetic_LessThan(con, -con.set.upper)
+end
+function _calculate_tight_M(
+    con::JuMP.ScalarConstraint{JuMP.AffExpr, <: _MOI.GreaterThan}
+    )
+    return _interval_arithmetic_GreaterThan(con, -con.set.lower)
+end
+function _calculate_tight_M(
+    con::JuMP.ScalarConstraint{JuMP.AffExpr, <: _MOI.Interval}
+    )
+    return [
+        _interval_arithmetic_GreaterThan(con, -con.set.lower),
+        _interval_arithmetic_LessThan(con, -con.set.upper)
+    ]
+end
+function _calculate_tight_M(
+    con::JuMP.ScalarConstraint{JuMP.AffExpr, <: _MOI.EqualTo}
+    )
+    return [
+        _interval_arithmetic_GreaterThan(con, -con.set.value),
+        _interval_arithmetic_LessThan(con, -con.set.value)
+    ]
+end
+_calculate_tight_M(con::JuMP.ScalarConstraint) = Inf
 
-    big_m_reformulation!(constr::NonlinearConstraintRef, bin_var, M, i, j, k)
+"""
 
-Perform Big-M reformulaiton on a nonlinear constraint at index k of constraint j in disjunct i.
+"""
+function _interval_arithmetic_LessThan(con::JuMP.ScalarConstraint, M::Float64)
+    for (var,coeff) in con.func.terms
```

Comment:
Sounds, good. I have added the `AffExpr` part to it. We cannot use interval arithmetic with the variable bounds to find the maximum and minimum of a quadratic expression because nonlinear expressions need you to solve a max or min optimization problem to find the largest or smallest value the expression can take. For quadratic expressions we could take the derivative and find the zero if we wanted to support infering the tightest big M for them.

Reply:
Ya for quadratics, I imagine there is an analytical formula we can code in.